src/toil/lib/docker.py

```diff
@@ -1,280 +1,340 @@
-                        subprocess.check_call(['docker', 'stop', containerName])
-            else:
-                _logger.info('The container "%s" was not found to be running.', containerName)
-            if action >= RM:
-                # If the container was run with --rm, then stop will most likely remove the
-                # container.  We first check if it is running then remove it.
-                running = _containerIsRunning(containerName)
-                if running is not None:
-                    _logger.info('Removing container "%s".', containerName)
-                    for attempt in retry(predicate=dockerPredicate):
-                        with attempt:
-                            subprocess.check_call(['docker', 'rm', '-f', containerName])
-                else:
-                    _logger.info('The container "%s" was not found on the system.  Nothing to remove.',
-                                 containerName)
-
-def _fixPermissions(tool, workDir):
+    try:
+        this_container = client.containers.get(container_name)
+        while this_container.status == 'running':
+            client.containers.get(container_name).kill()
+            this_container = client.containers.get(container_name)
+    except NotFound:
+        _logger.debug("Attempted to stop container, but container != exist: ",
+                      container_name)
+    except requests.exceptions.HTTPError as e:
+        _logger.debug("Attempted to stop container, but server gave an error: ",
+                      container_name)
+        raise create_api_error_from_http_exception(e)
+
+def _dockerStop(container_name, client):
     """
-    Fix permission of a mounted Docker directory by reusing the tool to change ownership.
-    Docker natively runs as a root inside the container, and files written to the
-    mounted directory are implicitly owned by root.
+    Gracefully kills a container.  Equivalent to "docker stop":
+    https://docs.docker.com/engine/reference/commandline/stop/
 
-    :param list baseDockerCall: Docker run parameters
-    :param str tool: Name of tool
-    :param str workDir: Path of work directory to recursively chown
+    :param container_name: Name of the container being stopped.
+    :param client: The docker API client object to call.
     """
-    if os.geteuid() == 0:
-        # we're running as root so this chown is redundant
-        return
-
-    baseDockerCall = ['docker', 'run', '--log-driver=none',
-                      '-v', os.path.abspath(workDir) + ':/data', '--rm', '--entrypoint=chown']
-    stat = os.stat(workDir)
-    command = baseDockerCall + [tool] + ['-R', '{}:{}'.format(stat.st_uid, stat.st_gid), '/data']
-    for attempt in retry(predicate=dockerPredicate):
-        with attempt:
-            subprocess.check_call(command)
-
-
-def _getContainerName(job):
-    return '--'.join([str(job),
-                      base64.b64encode(os.urandom(9), '-_')]).replace("'", '').replace('_', '')
-
+    try:
```

Comment:
This block of code is repeated in `_dockerKill`,  factor out?

Reply:
Updated.